### ISSUE TITLE
Possible FAT12 problem

### ISSUE THREAD

**Issue by Mellvik** [Mellvik/TLVC#47]
When mounting a 10MB FAT12 file system (on an ancient 10MB MFM drive), the `ts=` value reported on the console looks suspicious:
```
FAT: me=f8,csz=8,#f=2,floc=1,fsz=8,rloc=17,#d=512,dloc=49,#s=20723,ts=-1006632960
FAT: total 10M, fat12 format
```

The functionality seems unaffected but testing has been limited by driver related issues.
BTW, ELKS reports the exact same numbers.

**Comment by ghaerr** [comment]
This is likely something to do with the Total Sectors field of the DOS BPB stored in the drive's boot sector. Given the age of the drive, its possible our interpretation of the field doesn't account for something. IIRC the total sectors field was added in later versions of the evolving BPB. Out of curiosity, is DOS on the volume? What version?

This can be debugged more fully by running `hd` for the first sector of the drive so that we can analyze the BPB.

**Comment by Mellvik** [comment]
Thanks @ghaerr - that makes sense. 

There is DOS on the drive, but the keyboard and video are both broken (for now), making DOS inaccessible.

Also, this is a MFM drive which TLVC doesn't have a driver for yet, and BIOSHD on TLVC is mysteriously broken, so TLVC doesn't have disk access for now. The 360k diskette hasn't allowed for the `hd` command yet.

ELKS with BIOSHD works, but for some reason networking (wd8003) isn't working on XT ELKS (but works fine on TLVC) , so I don't have a means to populate a `bin` directory on the HD yet.

And interesting combination of missing pieces - IOW work in progress - including an interrupt/DMA TLVC driver for MFM drives.

**Comment by Mellvik** [comment]
The BPB looks like this:
```
000000: 33 c0 8e d8 8e c0 fa 8e  d0 bc 00 7c fb 8b f4 bf  3..........|....
000010: 00 06 b9 00 01 fc f3 a5  50 b8 1e 06 50 cb b9 04  ........P...P...
000020: 00 be be 07 80 3c 80 74  2c 83 c6 10 e2 f6 b4 0f  .....<.t,.......
000030: cd 10 b3 07 be 97 06 b9  78 00 ac b4 0e cd 10 e2  ........x.......
000040: f9 c7 06 72 04 34 12 b8  ff ff 50 1e b4 0f cd 10  ...r.4....P.....
000050: 32 e4 cd 16 cb b2 80 8b  dc 8a 74 01 bd 05 00 8b  2.........t.....
000060: 4c 02 b8 01 02 cd 13 73  1e 80 fc 11 75 10 1e b8  L......s....u...
000070: 00 f0 8e d8 33 ff 81 7d  ea 43 4f 1f 74 09 32 e4  ....3..}.CO.t.2.
000080: cd 13 4d 75 dd eb a7 81  bf fe 01 55 aa 75 9f 8b  ..Mu.......U.u..
000090: ee 1e 53 cb 43 52 4a 0d  0a 45 72 72 6f 72 20 6c  ..S.CRJ..Error l
0000a0: 6f 61 64 69 6e 67 20 6f  70 65 72 61 74 69 6e 67  oading operating
0000b0: 20 73 79 73 74 65 6d 20  66 72 6f 6d 20 66 69 78   system from fix
0000c0: 65 64 20 64 69 73 6b 2e  0d 0a 0d 0a 49 6e 73 65  ed disk.....Inse
0000d0: 72 74 20 43 4f 4d 50 41  51 20 44 4f 53 20 64 69  rt COMPAQ DOS di
0000e0: 73 6b 65 74 74 65 20 69  6e 20 64 72 69 76 65 20  skette in drive 
0000f0: 41 2e 0d 0a 50 72 65 73  73 20 61 6e 79 20 6b 65  A...Press any ke
000100: 79 20 77 68 65 6e 20 72  65 61 64 79 3a 20 07 00  y when ready: ..
000110: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  ................
*
0001e0: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 80 01  ................
0001f0: 01 00 01 03 51 30 11 00  00 00 f3 50 00 00 55 aa  ....Q0.....P..U.
```

**Comment by ghaerr** [comment]
>  debugged more fully by running hd for the first sector of the drive
> The BPB looks like this:

Sorry - the first sector of the HD is the MBR, as evidenced by the 0x80 at position 0x01ee, which shows an active partition number 4. We'll need the 1st sector of the *partition* instead to see the BPB. We can find that CHS sector number by decoding the MBR partition 4 [manually](https://en.wikipedia.org/wiki/Master_boot_record), or likely easier by running `fdisk` on the volume.

If both the above are too hard give the non-availability of tools, the first sector of a partition is usually the sector following the MBR (i.e. the second sector), or more likely, the first sector on the next track or cylinder following the MBR. That information is in the partition table.

A quick (untested) look at this MBR partition 4 entry above shows "01 01 00" as the start HSC (head/sector/cylinder), that is CHS 0/1/1. So it seems that the start of the partition is the first sector of cylinder 0 but the next track (head 1). Thus, adding the number of sectors per track for the HD will give us the sector offset for the start sector. Note this could also be verified by just dumping SPT+1 sectors of data, and checking the OEM indentifier in bytes 3-11 of the start sector BPB, which is usually something like "MSDOS 3.0" or similar.

**Comment by ghaerr** [comment]
Looking further into this, the MBR partition 4 shows the [partition type](https://en.wikipedia.org/wiki/Partition_type) as `01` which was used for DOS 2.0 media. DOS [BPB versions 2.0 through 3.2](https://en.wikipedia.org/wiki/BIOS_parameter_block) used a BPB that did not have the later 4-byte "Total Sectors" field (at offset 32) used starting with DOS 3.31 for FAT32 media.

All FAT12/16 filesystems use the original total sectors field (at offset 19), but starting in DOS 3.31 and later, the newer 4-byte Total Sectors field (at offset 32) used for FAT32 was set to zero for FAT12/16.

ELKS and TLVC display the `ts=` value from the 4-bytes at offset 32. So this isn't a bug, but rather for sure the issue is that your hard drive is formatted using a pre-DOS 3.31 BPB, something we've never seen before. Thus the value displayed is garbage.

Due to the partition type not being in the BPB but only in the MBR, as well as inconsistencies in accurately identifying a BPB version, it probably doesn't make much sense to try to not display the `ts=` value for FAT12/16, as it is useful for validating later BPB FAT mounts. It might be a good idea to change the `printk` format from %ld to %lu though.

**Comment by Mellvik** [comment]
Sorry - my mistake. `hd` replaced some lesser used binaries on the 360k image and is fully operational ...

I agree that since this field is informational only, AND that the occurence of such file systems is akin to a solar eclipse - it doesn't make much sense to add code (waste ram) to improve or fix (except the suggested `printk` change).

FWIW:
```
000200: eb 21 90 49 42 4d 20 20  33 2e 31 00 02 08 01 00  .!.IBM  3.1.....
000210: 02 00 02 03 51 f8 08 00  11 00 04 00 01 00 00 00  ....Q...........
000220: 00 00 00 c4 5c 08 33 ed  b8 c0 07 8e d8 33 c9 0a  ....\.3......3..
000230: d2 79 0e 89 1e 1e 00 8c  06 20 00 88 16 22 00 b1  .y....... ..."..
000240: 02 8e c5 8e d5 bc 00 7c  51 fc 1e 36 c5 36 78 00  .......|Q..6.6x.
000250: bf 23 7c b9 0b 00 f3 a4  1f 88 0e 2c 00 a0 18 00  .#|........,....
000260: a2 27 00 bf 78 00 b8 23  7c ab 91 ab a1 16 00 d1  .'..x..#|.......
000270: e0 40 e8 80 00 e8 86 00  bb 00 05 53 b0 01 e8 ab  .@.........S....
000280: 00 5f be 73 01 b9 0b 00  90 f3 a6 75 62 83 c7 15  ._.s.......ub...
000290: b1 0b 90 90 f3 a6 75 57  26 8b 47 1c 99 8b 0e 0b  ......uW&.G.....
0002a0: 00 03 c1 48 f7 f1 80 3e  71 01 60 75 02 b0 14 96  ...H...>q.`u....
0002b0: a1 11 00 b1 04 d3 e8 e8  3b 00 ff 36 1e 00 c4 1e  ........;..6....
0002c0: 6f 01 e8 39 00 e8 64 00  2b f0 76 0d e8 26 00 52  o..9..d.+.v..&.R
0002d0: f7 26 0b 00 03 d8 5a eb  e9 cd 11 b9 02 00 d3 e0  .&....Z.........
0002e0: 80 e4 03 74 04 fe c4 8a  cc 5b 58 ff 2e 6f 01 be  ...t.....[X..o..
0002f0: 89 01 eb 55 90 01 06 1e  00 11 2e 20 00 c3 a1 18  ...U....... ....
000300: 00 f6 26 1a 00 91 a1 1e  00 8b 16 20 00 f7 f1 92  ..&........ ....
000310: 8b 0e 18 00 f6 f1 2a cc  91 fe c5 86 e9 d0 ce d0  ......*.........
000320: ce 0a f1 86 f2 87 ca 8a  16 22 00 c3 98 bf 05 00  ........."......
000330: 50 b4 02 cd 13 72 02 58  c3 80 fc 11 74 f9 33 c0  P....r.X....t.3.
000340: cd 13 4f 58 75 ea be d3  01 e8 1d 00 be a9 01 e8  ..OXu...........
000350: 17 00 33 c0 cd 16 36 c7  06 72 04 34 12 ea 00 00  ..3...6..r.4....
000360: ff ff b4 0e bb 07 00 cd  10 ac 3c 24 75 f4 c3 00  ..........<$u...
000370: 00 70 00 49 42 4d 42 49  4f 20 20 43 4f 4d 49 42  .p.IBMBIO  COMIB
000380: 4d 44 4f 53 20 20 43 4f  4d 0a 0d 4e 6f 6e 2d 53  MDOS  COM..Non-S
000390: 79 73 74 65 6d 20 64 69  73 6b 20 6f 72 20 64 69  ystem disk or di
0003a0: 73 6b 20 65 72 72 6f 72  24 0a 0d 52 65 70 6c 61  sk error$..Repla
0003b0: 63 65 20 61 6e 64 20 73  74 72 69 6b 65 20 61 6e  ce and strike an
0003c0: 79 20 6b 65 79 20 77 68  65 6e 20 72 65 61 64 79  y key when ready
0003d0: 0a 0d 24 0a 0d 44 69 73  6b 20 62 6f 6f 74 20 66  ..$..Disk boot f
0003e0: 61 69 6c 75 72 65 24 43  6f 70 72 2e 20 43 4f 4d  ailure$Copr. COM
0003f0: 50 41 51 20 31 39 38 33  2c 38 34 2c 38 35 55 aa  PAQ 1983,84,85U.
```
`df`confirms what you observed from the MBR:
```
tlvc18# fdisk -l /dev/hda
                           START              END          SECTOR
Device           #:ID    Cyl Head Sect     Cyl Head Sect  Start   Size

/dev/hda1        1:00      0    0    0      0    0    0      0      0
/dev/hda2        2:00      0    0    0      0    0    0      0      0
/dev/hda3        3:00      0    0    0      0    0    0      0      0
/dev/hda4       *4:01      0    1    1     304    3   17     17  20723
tlvc18#
```

**Comment by Mellvik** [comment]
Fixed in #48.
